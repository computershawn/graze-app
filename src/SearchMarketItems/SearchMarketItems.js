import React, { Component } from 'react';
import './SearchMarketItems.css';
// In production, itemsListing info will
// be retrieved from the database 
import dummyStore from '../dummy-store'




class SearchMarketItems extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
      prevSearchTerm: '',
      haveResults: false,
      searchResults: [],
      itemNotFound: false
    }    
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(e) {
    let prevSearchTerm = this.state.searchTerm
    let searchResults = dummyStore.itemsListing.filter(item => {
      let foods = Object.keys(item.foods)
      return foods.includes(this.state.searchTerm.toLowerCase())
    })

    this.setState({
      haveResults: true,
      prevSearchTerm: prevSearchTerm,
      searchResults: searchResults,
      itemNotFound: searchResults.length === 0
    })

    e.preventDefault()
    // Insert code to find searchTerm in the database >>>
    // OR just retrieve all food items in database as JSON
    // and search for the search term in that data
  }

  handleChange(e) {
    const { name, value } = e.target
    this.setState({ [name]: value });
  }

  formatSearchResults() {
    const { prevSearchTerm } = this.state
    let resultsHTML = this.state.searchResults.map(result => {
      return (
        <article key={result.vendorName}>
          <h4>{result.marketName} | <a href="http://nothing.com">Map</a></h4>
          <p><em>
          ${result.foods[prevSearchTerm]} at {result.vendorName} (Stall {result.vendorStall})<br />
          </em></p>
        </article>
      )
    })
    return resultsHTML
  }

  render() {
    const { prevSearchTerm, haveResults, itemNotFound } = this.state
    return (
      <main>
        <header>
          <h1>Freshness Finder</h1>
        </header>

        <section>
          <form onSubmit={this.handleSubmit}>
            <div className="form-section">
              <label htmlFor="searchTerm">What are you shopping for today?</label>
              <input type="text" name="searchTerm" placeholder='Apples' onChange={this.handleChange} required />
            </div>
            <button type="submit">Find It!</button>
          </form>
        </section>

        {
          (haveResults && !itemNotFound) && (
            <section>
              <div className="results-title">
                <p>Search results for <strong>{prevSearchTerm}</strong>:</p>
              </div>
              <div>
                {this.formatSearchResults()}
              </div>
            </section>
        )}
        {
          itemNotFound && (
            <section>
              <div className="results-title">
                <p>Sorry, no markets in this area have <strong>{prevSearchTerm}</strong></p>
              </div>
            </section>
        )}
      </main>
    );
  }
}

export default SearchMarketItems;