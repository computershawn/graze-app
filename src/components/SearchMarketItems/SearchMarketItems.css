.search-field label,
.search-field input {
  font-size: 1.2em;
}

.search-field label {
  margin-bottom: 0.5em;
}

.search-container {
  margin-bottom: 2em;
}
.results-container {
  margin-bottom: 48px;
}

.results-area article {
  padding: 15px 0;
}

.results-area article p {
  text-indent: 0.5em;
}

.results-area a {
  text-decoration: none;
}

.results-area a:hover {
  text-decoration: underline;
}