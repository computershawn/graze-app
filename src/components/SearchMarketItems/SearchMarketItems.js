import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import Nav from '../Nav/Nav'
import './SearchMarketItems.css'
import MarketDataContext from '../../MarketDataContext'




class SearchMarketItems extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
      prevSearchTerm: '',
      haveResults: false,
      searchResults: [],
      itemNotFound: false
    }
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  static contextType = MarketDataContext;

  handleSubmit(e) {
    const { products=[], pricelist=[], vendors=[], markets=[] } = this.context;
    let prevSearchTerm = this.state.searchTerm
    let product = products.find(product => {
      return product.product_name === this.state.searchTerm.toLowerCase()
    })
    let index = product ? product.id : null
    let searchResults = []
    if(index) {
      let list = pricelist.filter(item => {
        return item.product_id.toString() === index.toString()
      })
      searchResults = list.map(item => {
        let p = item.price
        let u = item.units
        let v = vendors.find(vendor => {
          return vendor.id.toString() === item.vendor_id.toString()
        })
        let m = markets.find(market => {
          return market.id.toString() === v.market_id.toString()
        })
        return {
          price: `${p}/${u}`,
          marketId: m.id,
          market: m.market_name,
          vendor: v.vendor_name,
          stall: v.market_stall
        }
      })
    }

    this.setState({
      haveResults: true,
      prevSearchTerm: prevSearchTerm,
      searchResults: searchResults,
      itemNotFound: searchResults.length === 0
    })

    e.preventDefault()
  }

  handleChange(e) {
    const { name, value } = e.target
    this.setState({ [name]: value });
  }

  formatSearchResults() {
    let resultsHTML = this.state.searchResults.map((result, i) => {
      const linkTo = `/markets/${result.marketId}`
      return (
        <article key={i}>
          <h4>{result.market} | <a href="http://nothing.com">Map</a></h4>
          <p><em>
            ${result.price} at {result.vendor} (Stall {result.stall})<br />
          </em></p>
          <p><Link to={linkTo}>View this Market</Link></p>
        </article>
      )
    })
    return resultsHTML
  }

  render() {
    const { prevSearchTerm, haveResults, itemNotFound } = this.state
    return (
      <>
        <Nav />
        <main>
          <header>
            <h1>Freshness Finder</h1>
          </header>

          <section>
            <form onSubmit={this.handleSubmit}>
              <div className="form-section">
                <label htmlFor="searchTerm">What are you shopping for today?</label>
                <input type="text" name="searchTerm" placeholder='Apples' onChange={this.handleChange} required />
              </div>
              <button type="submit">Find It!</button>
            </form>
          </section>

          {
            (haveResults && !itemNotFound) && (
              <section>
                <div className="results-title">
                  <p>Search results for <strong>{prevSearchTerm}</strong></p>
                </div>
                <div>
                  {this.formatSearchResults()}
                </div>
              </section>
            )}
          {
            itemNotFound && (
              <section>
                <div className="results-title">
                  <p>Sorry, no markets in this area have <strong>{prevSearchTerm}</strong></p>
                </div>
              </section>
            )}
        </main>
      </>
    );
  }
}

export default SearchMarketItems;