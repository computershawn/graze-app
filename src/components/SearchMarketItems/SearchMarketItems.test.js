import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter } from 'react-router-dom'
import SearchMarketItems from './SearchMarketItems'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <BrowserRouter>
      <SearchMarketItems />
    </BrowserRouter>,
    div
  );
})

it('renders empty given no tabs', () => {
  const wrapper = shallow(<SearchMarketItems />)
  expect(toJson(wrapper)).toMatchSnapshot()
})