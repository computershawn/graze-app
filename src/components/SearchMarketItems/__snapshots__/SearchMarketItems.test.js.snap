// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty given no tabs 1`] = `
<Fragment>
  <Nav />
  <main>
    <header>
      <h1>
        Freshness Finder
      </h1>
    </header>
    <section>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-section"
        >
          <label
            htmlFor="searchTerm"
          >
            What are you shopping for today?
          </label>
          <input
            name="searchTerm"
            onChange={[Function]}
            placeholder="Apples"
            required={true}
            type="text"
          />
        </div>
        <button
          type="submit"
        >
          Find It!
        </button>
      </form>
    </section>
  </main>
</Fragment>
`;
